ansible_base/authentication/utils/authentication.py

```diff
@@ -0,0 +1,116 @@
+import logging
+from typing import Optional, Tuple
+
+from django.contrib.auth import get_user_model
+from django.contrib.auth.models import AbstractUser
+from social_core.pipeline.user import get_username
+
+from ansible_base.authentication.models import Authenticator, AuthenticatorUser
+from ansible_base.authentication.social_auth import AuthenticatorStorage, AuthenticatorStrategy
+
+logger = logging.getLogger('ansible_base.authentication.utils.authentication')
+
+
+class FakeBackend:
+    def setting(self, *args, **kwargs):
+        return ["username", "email"]
+
+
+def get_local_username(user_details: dict) -> str:
+    """
+    Converts the username provided by the backend to one that doesn't conflict with users
+    from other auth backends.
+    """
+    username = get_username(strategy=AuthenticatorStrategy(AuthenticatorStorage()), details=user_details, backend=FakeBackend())
+
+    return username['username']
+
+
+def determine_username_from_uid(uid: str, authenticator: Authenticator) -> str:
+    """
+    Determine what the username for the User object will be from the given uid and authenticator
+    This will take uid like "bella" and search for an AuthenticatorUser and return:
+        bella - if there is no bella user in AuthenticatorUser
+        bella<hash> - if there is already a bella user in AuthenticatorUser but its not from the given authenticator
+        <User.username> - If there is already a user associated with bella for this authenticator (could be bella or bella<hash> or even something else)
+    """
+    # If we have an AuthenticatorUser with the exact uid and provider than we have a match
+    exact_match = AuthenticatorUser.objects.filter(uid=uid, provider=authenticator)
+    if exact_match.count() == 1:
+        new_username = exact_match[0].user.username
+        logger.info(f"Authenticator {authenticator.name} already authenticated {uid} as {new_username}")
+        return new_username
+
+    # We didn't get an exact match. If any other provider is using this uid our id will be uid<hash>
+    if AuthenticatorUser.objects.filter(uid=uid).count() != 0:
+        # Some other provider is providing this username so we need to create our own username
+        new_username = get_local_username({'username': uid})
+        logger.info(
+            f'Authenticator {authenticator.name} wants to authenticate {uid} but that'
+            f'username is already in use by another authenticator,'
+            f' the user from this authenticator will be {new_username}'
+        )
+        return new_username
+
+    # We didn't have an exact match but no other provider is servicing this uid so lets return that for usage
+    logger.info(f"Authenticator {authenticator.name} is is able to authenticate user {uid} as {uid}")
+    return uid
+
+
+def get_or_create_authenticator_user(
+    username: str, authenticator: Authenticator, user_details: dict = dict(), extra_data: dict = dict()
```

Comment:
Replacing dict literal `{}` with a dict call `dict()` doesn't change anything. The dictionary object is still created at the function definition time. The common pattern in Python for creating mutable function defaults is using an immutable default with a condition. For example:

```
def foo(data: Optional[dict] = None):
    if data is None:
        data = {}
```

Reply:
In this case you should remove the default values. The `user_detauls: dict = dict` is incorrect. This assigns `dict` type as a default value, which I think not what was intended here.

Reply:
Default removed.